web/app/js/components/Octopus.jsx

```diff
@@ -35,47 +35,55 @@ ArrowCol.propTypes = PropTypes.bool.isRequired;
 export default class Octopus extends React.Component {
   static defaultProps = {
     neighbors: {},
-    resource: {}
+    resource: {},
+    unmeshedSources: []
   }
 
   static propTypes = {
     neighbors: PropTypes.shape({}),
-    resource: PropTypes.shape({})
+    resource: PropTypes.shape({}),
+    unmeshedSources: PropTypes.arrayOf(PropTypes.shape({})),
   }
 
-  renderResourceSummary(resource, type) {
+  renderResourceSummary(resource, type, unmeshed) {
     return (
       <div key={resource.name} className={`octopus-body ${type}`}>
         <div className={`octopus-title ${type}`}>
           <this.props.api.ResourceLink
             resource={resource}
             linkText={displayName(resource)} />
         </div>
-        <div className="octopus-sr-gauge">
-          <Progress
-            className={getSrClassification(resource.successRate)}
-            type="dashboard"
-            format={() => metricToFormatter["SUCCESS_RATE"](resource.successRate)}
-            width={type === "main" ? 132 : 64}
-            percent={resource.successRate * 100}
-            gapDegree={180} />
-        </div>
-        <Metric title="RPS" value={metricToFormatter["REQUEST_RATE"](resource.requestRate)} />
-        <Metric title="P99" value={metricToFormatter["LATENCY"](_.get(resource, "latency.P99"))} />
+        {
+          unmeshed ? <div>Unmeshed</div> :
+          <div>
+            <div className="octopus-sr-gauge">
+              <Progress
+                className={getSrClassification(resource.successRate)}
+                type="dashboard"
+                format={() => metricToFormatter["SUCCESS_RATE"](resource.successRate)}
+                width={type === "main" ? 132 : 64}
+                percent={resource.successRate * 100}
+                gapDegree={180} />
+            </div>
+            <Metric title="RPS" value={metricToFormatter["REQUEST_RATE"](resource.requestRate)} />
+            <Metric title="P99" value={metricToFormatter["LATENCY"](_.get(resource, "latency.P99"))} />
+          </div>
+        }
       </div>
     );
   }
 
   render() {
-    let { resource, neighbors } = this.props;
-    let hasUpstreams = _.size(neighbors.upstream) > 0;
+    let { resource, neighbors, unmeshedSources } = this.props;
+    let hasUpstreams = _.size(neighbors.upstream) > 0 || _.size(unmeshedSources) > 0;
     let hasDownstreams = _.size(neighbors.downstream) > 0;
 
     return (
       <div className="octopus-graph">
         <Row type="flex" justify="center" gutter={32} align="middle">
           <Col span={6} className={`octopus-col ${hasUpstreams ? "resource-col" : ""}`}>
-            {_.map(neighbors.upstream, n => this.renderResourceSummary(n, "neighbor"))}
+            {_.map(_.sortBy(neighbors.upstream, "resource.name"), n => this.renderResourceSummary(n, "neighbor"))}
+            {_.map(_.sortBy(unmeshedSources, "name"), n => this.renderResourceSummary(n, "neighbor", true))}
```

Comment:
I think I'll address these two issues in the same place. I've moved `updateNeighbors` to `indexTopResult` so the number of times it's called should be reduced by a factor of 3. It also uses fields from the already-parsed data, rather than re-parsing the entire raw payload.

Here, I also added `delete resourceAgg[key]` which will delete a resource from the `unmeshedResources` object if it becomes meshed.

Hopefully this will address some of the page sluggishness.

Reply:
o.o yeah... that hasn't fixed it

![screen shot 2018-09-06 at 4 04 55 pm](https://user-images.githubusercontent.com/549258/45189619-bca40a00-b1ee-11e8-9080-2eb4fa380bfb.png)